programs/drift/src/controller/liquidation.rs

```diff
@@ -648,6 +654,464 @@ pub fn liquidate_perp(
+        worst_case_base_asset_amount.unsigned_abs(),
+        MarginRequirementType::Maintenance,
+    )?;
+
+    let margin_ratio_with_buffer = margin_ratio.safe_add(liquidation_margin_buffer_ratio)?;
+
+    let margin_shortage = intermediate_margin_calculation.margin_shortage()?;
+
+    let market = perp_market_map.get_ref(&market_index)?;
+    let quote_spot_market = spot_market_map.get_ref(&market.quote_spot_market_index)?;
+    let quote_oracle_price = oracle_map.get_price_data(&quote_spot_market.oracle)?.price;
+    let liquidator_fee = market.liquidator_fee;
+    let if_liquidation_fee = calculate_perp_if_fee(
+        intermediate_margin_calculation.tracked_market_margin_shortage(margin_shortage)?,
+        user_base_asset_amount,
+        margin_ratio_with_buffer,
+        liquidator_fee,
+        oracle_price,
+        quote_oracle_price,
+        market.if_liquidation_fee,
+    )?;
+    let base_asset_amount_to_cover_margin_shortage = standardize_base_asset_amount_ceil(
+        calculate_base_asset_amount_to_cover_margin_shortage(
+            margin_shortage,
+            margin_ratio_with_buffer,
+            liquidator_fee,
+            if_liquidation_fee,
+            oracle_price,
+            quote_oracle_price,
+        )?,
+        market.amm.order_step_size,
+    )?;
+    drop(market);
+    drop(quote_spot_market);
+
+    let max_pct_allowed = calculate_max_pct_to_liquidate(
+        &user,
+        margin_shortage,
+        slot,
+        initial_pct_to_liquidate,
+        liquidation_duration,
+    )?;
+    let max_base_asset_amount_allowed_to_be_transferred =
+        base_asset_amount_to_cover_margin_shortage
+            .cast::<u128>()?
+            .saturating_mul(max_pct_allowed)
+            .safe_div(LIQUIDATION_PCT_PRECISION)?
+            .cast::<u64>()?;
+
+    if max_base_asset_amount_allowed_to_be_transferred == 0 {
+        msg!("max_base_asset_amount_allowed_to_be_transferred == 0");
+        return Ok(());
+    }
+
+    let base_asset_value =
+        calculate_base_asset_value_with_oracle_price(user_base_asset_amount.cast()?, oracle_price)?
+            .cast::<u64>()?;
+
+    // if position is less than $50, liquidator can liq all of it
+    let min_base_asset_amount = if base_asset_value > 50 * QUOTE_PRECISION_U64 {
+        0_u64
+    } else {
+        user_base_asset_amount
+    };
+
+    let base_asset_amount = user_base_asset_amount
+        .min(max_base_asset_amount_allowed_to_be_transferred.max(min_base_asset_amount));
+    let base_asset_amount = standardize_base_asset_amount_ceil(
+        base_asset_amount,
+        perp_market_map.get_ref(&market_index)?.amm.order_step_size,
+    )?;
+
+    let existing_direction = user.perp_positions[position_index].get_direction();
+
+    let order_params = get_liquidation_order_params(
+        market_index,
+        existing_direction,
+        base_asset_amount,
+        oracle_price,
```

Comment:
Is it possible for this exhibit some auction behavior instead of deriving an order based on `oracle_price`? Something like "if the user has been in liquidation for x slots, make the order more aggressive by y bps (or in price terms)"

Calculating the effective liq price on `oracle_price + liquidator_fee` has the downside of favoring short liqs (buy to close) when `oracle_price > drift_mark` and favoring long liqs when `oracle_price < drift_mark`. Which is most problematic if all makers and the AMM are quoting wider (or AMM is unavailable) than `oracle_price + liquidator_fee`.

Reply:
think this is a good follow up todo after merging this version